packages/instrumentation-openai/src/instrumentation.ts

```diff
@@ -0,0 +1,852 @@
+              result as ChatCompletion,
+              config,
+              ctx
+            );
+          })
+          .catch(
+            self._createAPIPromiseRejectionHandler(startNow, span, commonAttrs)
+          );
+
+        return apiPromise;
+      };
+    };
+  }
+
+  /**
+   * Start a span for this chat-completion API call. This also emits log events
+   * as appropriate for the request params.
+   */
+  _startChatCompletionsSpan(
+    params: ChatCompletionCreateParams,
+    config: OpenAIInstrumentationConfig,
+    baseURL: string | undefined
+  ) {
+    // Attributes common to span, metrics, log events.
+    const commonAttrs: Attributes = {
+      [ATTR_GEN_AI_OPERATION_NAME]: 'chat',
+      [ATTR_GEN_AI_REQUEST_MODEL]: params.model,
+      [ATTR_GEN_AI_SYSTEM]: 'openai',
+    };
+    Object.assign(commonAttrs, getAttrsFromBaseURL(baseURL, this._diag));
+
+    // Span attributes.
+    const attrs: Attributes = {
+      ...commonAttrs,
+    };
+    if (params.frequency_penalty != null) {
+      attrs[ATTR_GEN_AI_REQUEST_FREQUENCY_PENALTY] = params.frequency_penalty;
+    }
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    if ((params as any).max_completion_tokens != null) {
+      attrs[ATTR_GEN_AI_REQUEST_MAX_TOKENS] =
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        (params as any).max_completion_tokens;
+    } else if (params.max_tokens != null) {
+      // `max_tokens` is deprecated in favour of `max_completion_tokens`.
+      attrs[ATTR_GEN_AI_REQUEST_MAX_TOKENS] = params.max_tokens;
+    }
+    if (params.presence_penalty != null) {
+      attrs[ATTR_GEN_AI_REQUEST_PRESENCE_PENALTY] = params.presence_penalty;
+    }
+    if (params.stop != null) {
+      if (Array.isArray(params.stop)) {
+        attrs[ATTR_GEN_AI_REQUEST_STOP_SEQUENCES] = params.stop;
+      } else {
+        attrs[ATTR_GEN_AI_REQUEST_STOP_SEQUENCES] = [params.stop];
+      }
+    }
+    if (params.temperature != null) {
+      attrs[ATTR_GEN_AI_REQUEST_TEMPERATURE] = params.temperature;
+    }
+    if (params.top_p != null) {
+      attrs[ATTR_GEN_AI_REQUEST_TOP_P] = params.top_p;
+    }
+
+    const span: Span = this.tracer.startSpan(
+      `${attrs[ATTR_GEN_AI_OPERATION_NAME]} ${attrs[ATTR_GEN_AI_REQUEST_MODEL]}`,
+      {
+        kind: SpanKind.CLIENT,
+        attributes: attrs,
+      }
+    );
+    const ctx: Context = trace.setSpan(context.active(), span);
+
+    // Capture prompts as log events.
+    const timestamp = Date.now();
+    params.messages.forEach((msg: ChatCompletionMessageParam) => {
+      switch (msg.role) {
+        case 'system': {
+          const body: GenAISystemMessageEventBody = {};
+          if (config.captureMessageContent) {
+            if (Array.isArray(msg.content)) {
+              body.content = msg.content.map(p => p.text).join('');
+            } else {
+              body.content = msg.content;
+            }
+          }
+          this.logger.emit({
```

Comment:
If one has set up a global logging provider, is it possible to opt out of the logs emitted from this instrumentation?

Reply:
@seemk I think one could do so by setting a NoopLoggerProvider on the instrumentation via `instr.setLoggerProvider(noopLoggerProvider)`.

But are you asking for something different than that?

I haven't been following the GenAI semconv discussions of late. I vaguely recall that there was discussion of making it user-optional whether GenAI instrumentations used log-events or span attributes. Is that what you are referring to?

Reply:
Yeah - from what I understand, the conventions imply that when logs are enabled, logs are always emitted - this is how it currently works in the bedrock instrumentation of this repo, as well as genai in python and java. If it seems like there should be more knobs, we'd need to raise changes to the semconv but for now this implements the current practice.